File: app/src/main/java/com/ignis/android_cleanarchitecture/presentation/view/fragment/MainFragment.java
package com.ignis.android_cleanarchitecture.presentation.view.fragment;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ignis.android_cleanarchitecture.R;
import com.ignis.android_cleanarchitecture.databinding.FragmentMainBinding;
import com.ignis.android_cleanarchitecture.presentation.listener.activity.MainActivityListener;
import com.ignis.android_cleanarchitecture.presentation.listener.fragment.MainFragmentListener;
import com.ignis.android_cleanarchitecture.presentation.presenter.fragment.MainFragmentViewModel;

/**
 * Main Fragment
 *
 * @author Kensuke
 */
public class MainFragment extends Fragment implements MainFragmentListener {

    private FragmentMainBinding binding;
    private MainFragmentViewModel viewModel;
    private MainActivityListener mainActivityListener;

    public static MainFragment newInstance() {
        MainFragment fragment = new MainFragment();
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
            mainActivityListener = (MainActivityListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement " + MainActivityListener.class.getSimpleName());
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if (binding == null) {
            binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false);
            viewModel = new MainFragmentViewModel(getActivity(), this);
            binding.setViewModel(viewModel);
        }
        viewModel.onCreateView(savedInstanceState);
        return binding.getRoot();
    }

    @Override
    public void onStart() {
        super.onStart();
        viewModel.onStart();
    }

    @Override
    public void onStop() {
        viewModel.onStop();
        super.onStop();
    }

    /* MainFragment.MainFragmentListener */

    @Override
    public void setActionBarTitle(String title) {
        mainActivityListener.setActionBarTitle(title);
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/presentation/view/fragment/MainFragment.kt
package com.kazakago.cleanarchitecture.presentation.view.fragment

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Toast
import com.kazakago.cleanarchitecture.R
import com.kazakago.cleanarchitecture.presentation.extension.formattedText
import com.kazakago.cleanarchitecture.presentation.listener.fragment.MainFragmentViewModelListener
import com.kazakago.cleanarchitecture.presentation.presenter.fragment.MainFragmentViewModel
import com.kazakago.cleanarchitecture.presentation.view.adapter.CitySpinnerAdapter
import com.kazakago.cleanarchitecture.presentation.view.adapter.ForecastRecyclerAdapter
import kotlinx.android.synthetic.main.fragment_main.*

class MainFragment : Fragment(), MainFragmentViewModelListener {

    interface Listener {
        fun setActionBarTitle(title: String?)
    }

    companion object {
        fun createInstance(): MainFragment = MainFragment()
    }

    private val viewModel by lazy { ViewModelProviders.of(this).get(MainFragmentViewModel::class.java) }
    private lateinit var citySpinnerAdapter: CitySpinnerAdapter
    private lateinit var forecastRecyclerAdapter: ForecastRecyclerAdapter
    private var listener: Listener? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = context as? Listener
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        lifecycle.addObserver(viewModel)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_main, container, false)
        viewModel.listener = this
        fragmentManager
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        citySpinnerAdapter = CitySpinnerAdapter(activity!!)
        citySpinner.adapter = citySpinnerAdapter

        forecastRecyclerAdapter = ForecastRecyclerAdapter(activity!!)
        forecastRecyclerAdapter.listener = viewModel
        forecastRecyclerView.adapter = forecastRecyclerAdapter

        viewModel.cityList.observe(this, Observer {
            citySpinnerAdapter.cityList = it ?: listOf()
            citySpinnerAdapter.notifyDataSetChanged()
        })
        viewModel.selectedPosition.observe(this, Observer {
            it?.let {
                citySpinner.setSelection(it)
                listener?.setActionBarTitle(citySpinnerAdapter.cityList[it].name)
            }
        })
        viewModel.weather.observe(this, Observer {
            areaTextView.text = it?.location?.area
            prefectureTextView.text = it?.location?.prefecture
            cityTextView.text = it?.location?.city
            publicTimeTextView.text = getString(R.string.public_time, it?.publicTime?.formattedText(context!!) ?: "")
            forecastRecyclerAdapter.forecastList = it?.forecasts ?: listOf()
            forecastRecyclerAdapter.notifyDataSetChanged()
        })
        refreshButton.setOnClickListener {
            viewModel.onClickRefresh()
        }
        citySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                viewModel.onCitySelected(position)
            }

            override fun onNothingSelected(parent: AdapterView<*>) {
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        viewModel.listener = null
    }

    /* MainFragmentViewModelListener */

    override fun showToast(message: String?) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

}
